os/20211014-s390x-secure-execution.md

```diff
@@ -0,0 +1,241 @@
+ - adds new `/se` partition for `sdboot`
+ - adds `roothash` and `boothash` partitions for [dm-verity](https://docs.kernel.org/admin-guide/device-mapper/verity.html) setup
+ - calculates hashes for `/boot` and `/` and appends them to kernel's cmdline
+ - generates `/se/sdboot` encrypted by universal public hostkey (valid for all Z15 and newer machines)
+ - `zipl`es `/se/sdboot`, so `qemu-secex.qcow2` image could only be run with Secure Execution
+
+Customer should add his public hostkeys(s) into `ignition` config.
+
+## Step 3: During firstboot `/boot` and `rootfs` are encrypted with randomly generated LUKS keys; `integrity` option is enabled
+
+During firstboot `/boot` and `/` are luks encrypted with randomly generated luks keys. This ensures, that each CoreOS isntance has its own LUKS keys and also makes it not possible to read the contents of qcow2 image just by inspecting (`libguestfish`) it. Any modifications of qcow image will be detected in runtime by [dm-integrity](https://docs.kernel.org/admin-guide/device-mapper/dm-integrity.html).
+
+
+After provisioning, new disk is like:
+```
+NAME         PARTLABEL TYPE  MOUNTPOINT
+vdb                    disk
+|-vdb1       se        part
+|-vdb3       boot      part
+| `-boot_dif           crypt
+|   `-boot             crypt /boot
+`-vdb4       root      part
+  `-root_dif           crypt
+    `-root             crypt /sysroot
+
+```
+
+- `se`   - partition for newly generated`sdboot` encrypted image
+- `boot` - LUKS encrypted boot partition
+- `root` - LUKS encrypted root partition
+- `boot_dif` - `dm-integrity` for `/boot`
+- `root_dif` - `dm-integrity` for `/root`
+
+## Step 4: Ignition protection
+
+Proposal:
+
+The user will encrypt the Ignition config with his symmetric key and the Ignition public key for the pre-built image.
+The private key is part of the `sdboot` image and is randomly generated during qcow2 image build. The public key is part of the build metadata. During firstboot the Ignition config gets decrypted.
+We should avoid printing the Ignition config to the console (e.g. serial and VGA) and scrub it before any shell access is possible.
+
+- In the SE path, add dropins to disable Ignition logging
+- Right after config decryption drop the ignition's private key to avoid leaking it.
+- In the emergency shell code, delete `/run/ignition.json` before entering the shell.
+- User generates the symmetric `key` and encrypts with it the `config.ign`:
+  ```
+  openssl rand -base64 96 > key
+  openssl enc -salt -pbkdf2 -aes256 -in config.ign -out config.crypted -pass file:./key
+  ```
+- User encryptes his `key` (using `ignition.pub`) and generates a tarball:
+  ```
+  openssl pkeyutl -encrypt -in key -out key.crypted -pubin -inkey /path/to/ignition.pub
+  find . -mindepth 1 -name "*.crypted" | tar -cvf /path/to/ignition_crypted -T -
```

Comment:
Doesn't this imply that the initramfs is unpacking an untrusted tar file?  That puts `tar` in a security-sensitive role, which doesn't seem ideal.  I'd strongly favor using existing hardened tools rather than assembling our own.

Reply:
tar here is only for convenience. we can add one more `virtio-blk` dev for a symmetric key